# Exploring eBay Car Sales Data

## Table of Content

- Introduction
- Cleaning Column Names
- Exploring the Data
- Data Cleaning
- Exploring Date Columns
- Exploring Price by Band
- Storing Aggregate Data in a DataFrame
- Translating German Words to English
- Damaged Cars vs Undamaged Cars
- Analysis by Brand/Model Combinations
- Conclusion



## Introduction

In this project we want to quantitatively analyze an included used car listings on a German eBay site to glean insights for a business that wants to operate in the automotive world.

### The Problem Significance

If a person who is interested to do a business in automotive industry decides to open a shop to sell used cars, it would be helpful to them to find answers to some of these questions:

- What are the top brands and their mean prices?
- Does the average price of the top brands have a correlation to the average mileage?
- How much cheaper are cars with damage than their non-damaged counterparts?
- What are the common brand/model combinations?
 
This study will extract insights from a popular dataset that fairly represents this industry to help anyone who wants to do this business make valuable decisions.

### Aim

The aim of this project is to clean the data and analyze the included used car listings.

### The dataset

In this project, we'll work with a dataset of used cars from *eBay Kleinanzeigen*, a [classified](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The data dictionary provided with data is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.

- `name` - Name of the car.

- `seller` - Whether the seller is private or a dealer.

- `offerType` - The type of listing

- `price` - The price on the ad to sell the car.

- `abtest` - Whether the listing is included in an A/B test.

- `vehicleType` - The vehicle Type.

- `yearOfRegistration` - The year in which the car was first registered.

- `gearbox` - The transmission type.

- `powerPS` - The power of the car in PS.

- `model` - The car model name.

- `odometer` - How many kilometers the car has driven.

- `monthOfRegistration` - The month in which the car was first registered.

- `fuelType` - What type of fuel the car uses.

- `brand` - The brand of the car.

- `notRepairedDamage` - If the car has a damage which is not yet repaired.

- `dateCreated` - The date on which the eBay listing was created.

- `nrOfPictures` - The number of pictures in the ad.

- `postalCode` - The postal code for the location of the vehicle.

- `lastSeenOnline` - When the crawler saw this ad last online.

### Source

The dataset was originally scraped and uploaded to Kaggle by user [orgesleka](https://www.kaggle.com/orgesleka).
The original dataset isn't available on Kaggle anymore, but you can find it [here](https://data.world/data-society/used-cars-data).


## Cleaning Column Names


In [1]:
# Importing Necessary Libraries and Reading of Dataset
import pandas as pd
import numpy as np

# Latin-1 encoding was used since we got an encoding error with utf-8
autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [2]:
# information about the autos dataframe and the first few rows
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We can make the following observations about the dataset:

- The dataset contains 50000 rows and 20 columns, most of which are string(object)
- Some columns have null values, but none has more than ~20% null values
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
# using dataframe.rename method to reword some of the column names based on the data dictionary to be more descriptive.
autos.rename(columns={
    'yearOfRegistration' : 'registration_year',
    'monthOfRegistration' : 'registration_month',
    'notRepairedDamage' : 'unrepaired_damage',
    'dateCreated' : 'ad_created'   
}, inplace = True)

In [5]:
# this function changes the camel case to snake case
import re

def camel_to_snake(name):
    name = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', name).lower()

new_column_names = []
for c in autos.columns:
    new_column_names.append(camel_to_snake(c))
    
autos.columns = new_column_names

In [6]:
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Exploring the Data

Let's explore the data and specifically look for data that doesn't look right.

In [8]:
autos.describe(include='all') #this line get both categorical and numeric columns

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

After looking at the basic statistics of our dataset more in depth, we make the following observations:

### Observation 1:
Rather than having numeric values, price and odometer are showing NaN values for numeric statistics (like mean, std etc.). If we run a datatype check for our dataframe we identify that these two columns have 'object' as datatype, which simply means that their values are strings. For statistical analysis purposes, we will clear non-numeric characters (if any) from the values in those two columns and convert the datatype of the columns to a numeric dataype 

In [9]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                 object
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer              object
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
nr_of_pictures         int64
postal_code            int64
last_seen             object
dtype: object

### Observation 2:

The `registration_year` column contains some odd values:

- The minimum value is 1000, before cars were invented
- The maximum value is 9999, many years into the future

## Data cleaning

Let's remove any non-numeric values from the price and odometer columns and convert the datatype from object to numeric datatype (int). 

In [10]:

autos['odometer'] = autos['odometer'].str.replace('(\D+)', '')
# The first parameter is the regular expression whose every occurrence was replaced with the second parameter’s value.
# we used \D, which is used to match all the characters that are not decimal digits.
autos['price'] = autos['price'].str.replace('(\D+)', '')

C:\Users\Kofi Adom\AppData\Local\Temp/ipykernel_26728/3152554717.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  autos['odometer'] = autos['odometer'].str.replace('(\D+)', '')
C:\Users\Kofi Adom\AppData\Local\Temp/ipykernel_26728/3152554717.py:4: FutureWarning: The default value of regex will change from True to False in a future version.
  autos['price'] = autos['price'].str.replace('(\D+)', '')


In [11]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type price   abtest vehicle_type  registration_year  \
0  privat    Angebot  5000  control          bus               2004   
1  privat    Angebot  8500  control    limousine               1997   
2  privat    Angebot  8990     test    limousine               2009   
3  privat    Angebot  4350  control   kleinwagen               2007   
4  privat    Angebot  1350     test        kombi               2003   

     gearbox  power_ps   model odometer  registration_month fuel_type  \
0    manuell       158  andere   150000                   3       lpg   
1  automatik       286     7er   150000                   6    benzin   
2    manuell       102    golf    70000                   7    benzin   
3  automatik        71  fortwo    70000                   6    benzin   
4    manuell         0   focus   150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [12]:
# convert the price and the odometer columns to numeric datatype
autos['price'] = autos['price'].astype(int)
autos['odometer'] = autos['odometer'].astype(int)

In [13]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                  int32
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer               int32
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
nr_of_pictures         int64
postal_code            int64
last_seen             object
dtype: object

We will rename the `odometer` column as `odometer_km` to reflect that the numbers are measured in km.

In [14]:
autos.rename(columns={'odometer': 'odometer_km'}, inplace=True)

In [15]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

Since we have the values for both price and `odometer_km` columns as numeric, let's continue exploring the data, specifically looking for data that doesn't look right.

We will start by exploring the `price` column

In [16]:
autos['price'].unique().shape

(2357,)

In [17]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [18]:
# prints the number of the first 20 unique values in ascending order
print(autos["price"].value_counts().sort_index(ascending = False).head(20))

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


After taking a closer look at the statistical analysis for the price column, we identify that the maximum value is way too high, and it's probably the reason why our standard deviation is large. There seems to be a steady increase in price until 350K after which a biggest jump in price occurs. In order to prevent our data from being skewed, we will remove rows where the price was over 350K.

In [19]:
autos = autos[autos["price"].between(0, 350000)]

print(autos["price"].describe())

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price, dtype: float64


Next, we will further explore the `odometers_km` column 

In [20]:
autos['odometer_km'].unique().shape

(13,)

In [21]:
autos['odometer_km'].describe()

count     49986.000000
mean     125736.506222
std       40038.133399
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [22]:
print(autos["odometer_km"].value_counts().sort_index(ascending = False).head(20))

150000    32416
125000     5169
100000     2168
90000      1757
80000      1436
70000      1230
60000      1164
50000      1025
40000       818
30000       789
20000       784
10000       264
5000        966
Name: odometer_km, dtype: int64


There are no outliers in the `odometer_km` column. Therefore we won't remove any row. We will keep on exploring the data and once we find other odd values we will clear or fix them.

## Exploring the date columns

There are 5 columns that represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

The `registration_month` and the `registration_year` columns are identified as inteher values but the values in the date_crawled, last_seen and ad_created columns are stored as strings. We need to convert the string data into a numerical representation so we can understand it quantitatively.

In [23]:
# checking how the values in the three string columns are formatted
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We notice that the first 10 characters represent the day (e.g. 2016-03-12). To understand the date range, we can extract just the date values from each column, and calculate the distribution of the values.

In [24]:
# selecting the first 10 characters (day part) in each column

date_crawled_days = (autos['date_crawled']
                         .str[:10]
                         # include missing values in the distribution and
                         # use percentage instead of counts
                         .value_counts(normalize=True, dropna=False)
                         .sort_index()

)

In [25]:
ad_created_days = (autos['ad_created']
                       .str[:10]
                       .value_counts(normalize=True, dropna=False)
                       .sort_index()
)

In [26]:
last_seen_days = (autos['last_seen']
                      .str[:10]
                      .value_counts(normalize=True, dropna=False)
                      .sort_index()

                     )

In [27]:
print(date_crawled_days)

2016-03-05    0.025387
2016-03-06    0.013944
2016-03-07    0.035970
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-10    0.032129
2016-03-11    0.032489
2016-03-12    0.036770
2016-03-13    0.015564
2016-03-14    0.036630
2016-03-15    0.033990
2016-03-16    0.029508
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-22    0.032909
2016-03-23    0.032389
2016-03-24    0.029108
2016-03-25    0.031749
2016-03-26    0.032489
2016-03-27    0.031049
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033629
2016-03-31    0.031909
2016-04-01    0.033809
2016-04-02    0.035410
2016-04-03    0.038691
2016-04-04    0.036490
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64


There is an even distribution of values in the `date_crawled` column. 

In [28]:
print(ad_created_days)

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
                ...   
2016-04-03    0.038931
2016-04-04    0.036850
2016-04-05    0.011843
2016-04-06    0.003261
2016-04-07    0.001280
Name: ad_created, Length: 76, dtype: float64


There are 76 unique days in the `ad_created` column

In [29]:
# let's print the first 50 values for a better understanding of the distribution

print(ad_created_days.head(50))

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
2016-02-22    0.000020
2016-02-23    0.000080
2016-02-24    0.000040
2016-02-25    0.000060
2016-02-26    0.000040
2016-02-27    0.000120
2016-02-28    0.000200
2016-02-29    0.000160
2016-03-01    0.000100
2016-03-02    0.000100
2016-03-03    0.000860
2016-03-04    0.001440
2016-03-05    0.023046
2016-03-06 

In [30]:
print(ad_created_days.tail(30))

2016-03-09    0.033229
2016-03-10    0.031869
2016-03-11    0.032789
2016-03-12    0.036610
2016-03-13    0.016925
2016-03-14    0.035230
2016-03-15    0.033749
2016-03-16    0.030008
2016-03-17    0.031189
2016-03-18    0.013724
2016-03-19    0.033849
2016-03-20    0.037871
2016-03-21    0.037691
2016-03-22    0.032769
2016-03-23    0.032189
2016-03-24    0.029088
2016-03-25    0.031889
2016-03-26    0.032569
2016-03-27    0.030909
2016-03-28    0.034970
2016-03-29    0.034110
2016-03-30    0.033449
2016-03-31    0.031909
2016-04-01    0.033809
2016-04-02    0.035090
2016-04-03    0.038931
2016-04-04    0.036850
2016-04-05    0.011843
2016-04-06    0.003261
2016-04-07    0.001280
Name: ad_created, dtype: float64


**March and April of 2016** seemed to be the most popular time to create adds.

In [31]:
print(last_seen_days)

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015884
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015744
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018585
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-26    0.016965
2016-03-27    0.016024
2016-03-28    0.020846
2016-03-29    0.022326
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023106
2016-04-02    0.024887
2016-04-03    0.025367
2016-04-04    0.024627
2016-04-05    0.124275
2016-04-06    0.220982
2016-04-07    0.130957
Name: last_seen, dtype: float64


There is an even distribution in the days for `last_seen` column.

Now we look at the distribution on the `registration_year` and the `registration_month` columns

In [32]:
autos[['registration_year', 'registration_month']].describe()

registration_year  registration_month
count       49986.000000        49986.000000
mean         2005.075721            5.723723
std           105.727161            3.711839
min          1000.000000            0.000000
25%          1999.000000            3.000000
50%          2003.000000            6.000000
75%          2008.000000            9.000000
max          9999.000000           12.000000

The `registration_year` column contains some odd values:

- The **minimum value** is 1000, before cars were invented
- The **maximum value** is 9999, many years into the future

The `registration_month` starts with 0, which is wrong

In [33]:
print(autos['registration_month'].value_counts().sort_index())

0     5073
1     3280
2     3006
3     5069
4     4101
5     4107
6     4367
7     3948
8     3191
9     3388
10    3651
11    3360
12    3445
Name: registration_month, dtype: int64


10% of the values in the `registration_month` column are invalid (we don't have month 0)

Now let's check the unique values in the `registration_year` column and find out how to fix the odd values

In [34]:
print(autos['registration_year']
                  .value_counts(dropna=False)
                  .sort_index(ascending=False)
                  .head(15)
)

9999       4
9996       1
9000       2
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       3
2018     491
2017    1452
2016    1316
Name: registration_year, dtype: int64


In [35]:
print(autos['registration_year']
                  .value_counts(dropna=False)
                  .sort_index(ascending=False)
                  .tail(15)
)

1943    1
1941    2
1939    1
1938    1
1937    4
1934    2
1931    1
1929    1
1927    1
1910    9
1800    2
1500    1
1111    1
1001    1
1000    1
Name: registration_year, dtype: int64


There are 5 unique values which are too early. They are years before 1866 (the year in which the first car was invented). 11 values are also invalid since it is impossible to list a car in 2016 that has been registered a year or more later. We will remove those rows and the rows which contain invalid month values.

In [36]:
autos = autos[autos['registration_year'].between(1866, 2016)]
autos = autos[autos['registration_month'].between(1, 12)]

In [37]:
print(autos['registration_year'].value_counts(normalize=True).sort_index())
# normalize=True gives the percentage of a value

1910    0.000023
1927    0.000023
1929    0.000023
1931    0.000023
1934    0.000023
          ...   
2012    0.030048
2013    0.018305
2014    0.015105
2015    0.008519
2016    0.022427
Name: registration_year, Length: 78, dtype: float64


In [38]:
print(autos['registration_month'].value_counts(normalize=True).sort_index())

1     0.072807
2     0.067096
3     0.112756
4     0.091020
5     0.090928
6     0.096961
7     0.087750
8     0.071034
9     0.075984
10    0.081833
11    0.075317
12    0.076514
Name: registration_month, dtype: float64


Now, the values in the registration_year and the registration_month are reasonable

## Exploring Price by Brand

When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the brand column. At this stage we want to proceed to analyze the car by brand and find out how is each brand priced. 

In [39]:
# print the counts for each band
print(autos['brand'].value_counts(dropna=False))

volkswagen        9060
bmw               4873
opel              4537
mercedes_benz     4308
audi              3822
ford              2973
renault           2035
peugeot           1295
fiat              1107
seat               796
skoda              730
nissan             668
mazda              649
smart              620
citroen            599
toyota             573
hyundai            440
sonstige_autos     414
volvo              404
mini               402
mitsubishi         345
honda              341
kia                328
alfa_romeo         285
porsche            277
suzuki             261
chevrolet          248
chrysler           158
dacia              118
daihatsu           104
jeep               101
land_rover          93
subaru              89
saab                73
jaguar              70
daewoo              65
rover               57
trabant             44
lancia              44
lada                24
Name: brand, dtype: int64


We want to select the brands which are at least 5% of the total values for aggregation.

In [40]:
brands_percentage = autos['brand'].value_counts(normalize=True)
print(brands_percentage)

volkswagen        0.208612
bmw               0.112204
opel              0.104467
mercedes_benz     0.099194
audi              0.088004
ford              0.068455
renault           0.046857
peugeot           0.029818
fiat              0.025489
seat              0.018328
skoda             0.016809
nissan            0.015381
mazda             0.014944
smart             0.014276
citroen           0.013792
toyota            0.013194
hyundai           0.010131
sonstige_autos    0.009533
volvo             0.009302
mini              0.009256
mitsubishi        0.007944
honda             0.007852
kia               0.007552
alfa_romeo        0.006562
porsche           0.006378
suzuki            0.006010
chevrolet         0.005710
chrysler          0.003638
dacia             0.002717
daihatsu          0.002395
jeep              0.002326
land_rover        0.002141
subaru            0.002049
saab              0.001681
jaguar            0.001612
daewoo            0.001497
rover             0.001312
t

In [41]:
# print the brands that make up at least 5% of the adds
brands_percentage = brands_percentage[brands_percentage >= 0.05]
print(brands_percentage)

volkswagen       0.208612
bmw              0.112204
opel             0.104467
mercedes_benz    0.099194
audi             0.088004
ford             0.068455
Name: brand, dtype: float64


There are 6 brands which have 5% or more than the total ads, with Volkswagen having the most and Ford, the least.

Now we will calculate the mean price per the 6 brands

In [42]:
# create a frequency table with the mean price for each brand
brands_prices_dict = {}
top_6_brands = brands_percentage

for b in top_6_brands.index:
    price = autos.loc[autos['brand'] == b, 'price']
    mean_price = round(price.mean(), 2)
    brands_prices_dict[b] = mean_price
    
print(sorted(brands_prices_dict.items(), key=lambda x:x[1]))

[('opel', 3093.87), ('ford', 3902.51), ('volkswagen', 5639.91), ('bmw', 8483.43), ('mercedes_benz', 8771.39), ('audi', 9604.79)]


While Volkswagen made up most of the ads on eBay, Audi is the most expensive car and Opel is the least expensive car. 

## Storing Aggregate Data in a DataFrame

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. 

In [43]:
# create a frequency table with the mean price for each brand
brands_mileage_dict = {}

for b in top_6_brands.index:
    mileage = autos.loc[autos['brand'] == b, 'odometer_km']
    mean_mileage = round(mileage.mean(), 2)
    brands_mileage_dict[b] = mean_mileage
    
print(sorted(brands_mileage_dict.items(), key=lambda x:x[1]))

[('ford', 124007.74), ('volkswagen', 128338.85), ('audi', 128784.67), ('opel', 128880.32), ('mercedes_benz', 130966.81), ('bmw', 132537.45)]


In order to compare the price and the mileage of each of the 6 brands, we need to convert both the price and mileage dictionares to a series, and later on put them inside a dataframe:

In [44]:
# Convert the dictionary to series
mean_price_series = pd.Series(brands_prices_dict)
mean_mileage_series = pd.Series(brands_mileage_dict)

print(mean_price_series.sort_values(ascending=False))
print('\n')
print(mean_mileage_series.sort_values(ascending=False))

audi             9604.79
mercedes_benz    8771.39
bmw              8483.43
volkswagen       5639.91
ford             3902.51
opel             3093.87
dtype: float64


bmw              132537.45
mercedes_benz    130966.81
opel             128880.32
audi             128784.67
volkswagen       128338.85
ford             124007.74
dtype: float64


In [45]:
# create a dataframe with the mean_price series
top_6_price_mileage = pd.DataFrame(mean_price_series, columns=["mean_price"])
# add mean_mileage series to the dataframe we just created
top_6_price_mileage["mean_mileage"] = mean_mileage_series
top_6_price_mileage

mean_price  mean_mileage
volkswagen        5639.91     128338.85
bmw               8483.43     132537.45
opel              3093.87     128880.32
mercedes_benz     8771.39     130966.81
audi              9604.79     128784.67
ford              3902.51     124007.74

Based on this result, we observe that:

- Mileage is almost similar in most of the cars

- There in no visible link between the mean price and mean millage for the top 6 brands. One may think that the most expensive car should have the least mileage. However, Audi which is the most expensive car has the fourth least mileage but Ford which is the least expensive car has the least mileage.

This implies that brand plays a bigger part than mileage in pricing

## Translating German Words to English

After inspecting the data in the columns, we can identify that the following columns have german words: `offer_type`, `vehicle_type`, `gearbox`, `unrepaired_damaged`, and `fuel_type`. We will translate them and map the values to their english counterparts.

In [46]:
# the unique values for columns with non-english words
print(autos['offer_type'].unique())
print(autos['vehicle_type'].unique())
print(autos['gearbox'].unique())
print(autos['unrepaired_damage'].unique())
print(autos['fuel_type'].unique())

['Angebot']
['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']
['manuell' 'automatik' nan]
['nein' nan 'ja']
['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']


We will use Google Translate to translate the german words to English and create mapping dictionary

In [47]:
offer = {"Angebot":"Offer","Gesuch":"Request"}
vehicles = {"bus": "bus",
            "limousine": "limousine",
            "kleinwagen": "small car",
            "kombi": "combi",
            "suv": "suv",
            "cabrio": "cabrio",
            "andere": "others"
                
}
gear = {"manuell": "manual", "automatik": "automatic"}
unrepaired = {"nein":"no", "ja":"yes"}
fuel = {"lpg":"lpg",
        "benzin":"petrol",
        "diesel":"diesel",
        "cng":"cng",
        "hybrin":"hybrid",
        "elektro":"electric",
        "andere":"others"
} 

In [48]:
# replace german with English translation
autos["offer_type"] = autos["offer_type"].map(offer)
autos["vehicle_type"] = autos["vehicle_type"].map(vehicles)
autos["gearbox"] = autos["gearbox"].map(gear)
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(unrepaired)
autos["fuel_type"] = autos["fuel_type"].map(fuel)

# prints unique values again
print(autos["offer_type"].unique())
print(autos["vehicle_type"].unique())
print(autos["gearbox"].unique())
print(autos["unrepaired_damage"].unique())
print(autos["fuel_type"].unique())

['Offer']
['bus' 'limousine' 'small car' 'combi' nan 'suv' 'cabrio' 'others']
['manual' 'automatic' nan]
['no' nan 'yes']
['lpg' 'petrol' 'diesel' nan 'cng' 'electric' 'others']


Now we have all german words in english

## Damaged Cars vs Non-damaged Cars

Now we will look at the mean price for damaged and undamaged cars. We will first separate the damaged and undamaged cars. We notice that there are some nan values in this column. This is probably due to cars that were never damaged and did not need repairs.

In [49]:
# seperates the damaged and undamaged cars
damaged = autos[autos["unrepaired_damage"] == "yes"]
non_damaged = autos[autos["unrepaired_damage"] != "yes"]
print(damaged.shape[0])
print(non_damaged.shape[0])

print(autos["unrepaired_damage"].shape[0])

4135
39295
43430


In [54]:
mean_damaged = round(damaged["price"].mean(), 2)
mean_nondamaged = round(non_damaged["price"].mean(), 2)
print("The mean price for damaged cars is", mean_damaged)
print("The mean price for repaired cars is", mean_nondamaged)

The mean price for damaged cars is 2315.32
The mean price for repaired cars is 6608.2


Now let's find the difference between the mean prices of damaged cars and non-damaged cars

In [51]:
print(round(mean_nondamaged - mean_damaged, 2))

4292.88


On average cars that were damaged and not repaired were about 4292.88 dollars cheaper than cars that were repaired or never damaged.

## Analysis by Brand/Model Combinations

Finally, let's find the most common brand/model combinations. In decision-making, a knowledge of this could help one come up with a best strategy just like the information we got from price/mileage combination.

In [52]:
autos_by_brandmodel = autos.groupby(["brand", "model"])["brand"].count()
pd.set_option('display.max_rows', 291) #this line let Jupyter notebook to show all the values
print(autos_by_brandmodel)

brand          model             
alfa_romeo     145                      3
               147                     79
               156                     78
               159                     29
               andere                  54
               spider                  32
audi           100                     49
               200                      1
               80                     165
               90                       6
               a1                      82
               a2                      43
               a3                     791
               a4                    1165
               a5                     126
               a6                     760
               a8                      68
               andere                 215
               q3                      28
               q5                      62
               q7                      40
               tt                     138
bmw            1er                    512


In [53]:
autos_by_brandmodel_max = autos_by_brandmodel.sort_values()
print(autos_by_brandmodel_max)

brand          model             
bmw            i3                       1
ford           b_max                    1
rover          rangerover               1
audi           200                      1
rover          discovery                1
dacia          andere                   2
rover          freelander               2
lancia         kappa                    2
daihatsu       charade                  3
               materia                  3
alfa_romeo     145                      3
volvo          v60                      3
lancia         delta                    4
lada           andere                   4
               samara                   4
land_rover     andere                   4
fiat           croma                    4
saab           9000                     5
dacia          lodgy                    5
smart          andere                   5
seat           exeo                     5
land_rover     range_rover_evoque       5
volkswagen     amarok                   6


The analysis above can give someone the following for a decision making:
- The counts by brand and model.
- The combinations of brands and models from less common to more common.

## Conclusion

In conclusion, these are some of the key findings from our analysis of the eBay Kleinanzeigen dataset:

- March and April of 2016 seemed to be the most popular time to create adds.
- The top six brands were volkswagen (`$`5639.91), bmw (`$`8483.43), opel (`$`3093.87), mercedes_benz (`$`8771.39), audi (`$`9604.79), and ford (`$`3902.51)
- There in no visible link between the mean price and mean millage for the top 6 brands.
- Golf is the most common model for volkswagen.
- Cars that were damaged and not repaired were about `$`4292.88 cheaper than cars that were repaired or never damaged. 